### PR TITLE
Improve how we handle kv iterator interface

### LINKED ISSUE #4164: Improve how kv iterator internal iterator state
KV iterator that handle internal iterator handle internal call to store that can be failed.
Calling Close or Err on invalid state of the iterator can cause an access violation.

### PR DESCRIPTION
**Pull request by nopcoder**
- Align how repository iterator use of MessageIterator handle when it is closed
- Align how tag iterator use of MessageIterator handle when it is closed
- kv/PartitionIterator Err should check itr closed flag and not interface
- Remove interface compare in store test

Close #4164

### CODE REVIEW

File: pkg/graveler/ref/commit_ordered_iterator.go [hunk 1]
```diff
@@ -113,14 +89,42 @@ func (i *KVOrderedCommitIterator) Value() *graveler.CommitRecord {
...
 		return i.it.Err()
 	}
-	return i.err
+	return nil
 }
 
 func (i *KVOrderedCommitIterator) Close() {
 	if i.it != nil {
 		i.it.Close()
+		i.it = nil
```
Comment (N-o-Z) [inline pkg/graveler/ref/commit_ordered_iterator.go:104]: We basically want that for all iterators, no?

Comment (nopcoder) [reply pkg/graveler/ref/commit_ordered_iterator.go:104]: The ones that holds internal pointer to implementation - I'll map the rest - I think I went over the pkg/kv in this pr.

**Review by N-o-Z — CHANGES_REQUESTED** [review]
Thanks for adding this code.
I would love to see some tests that cover the additional logic

**Comment by nopcoder** [comment]
@N-o-Z added specific tests that would fail before the above change.

File: pkg/graveler/ref/commit_ordered_iterator_test.go [hunk 2]
```diff
@@ -8,6 +8,12 @@ import (
 	"testing"
 	"time"
 
+	"github.com/treeverse/lakefs/pkg/kv"
+
```
Comment (N-o-Z) [inline pkg/graveler/ref/commit_ordered_iterator_test.go:12]: Fix imports

File: pkg/graveler/ref/repository_iterator_test.go [hunk 3]
```diff
@@ -5,6 +5,10 @@ import (
 	"testing"
 	"time"
 
+	"github.com/golang/mock/gomock"
+	"github.com/treeverse/lakefs/pkg/kv"
+	"github.com/treeverse/lakefs/pkg/kv/mock"
+
```
Comment (N-o-Z) [inline pkg/graveler/ref/repository_iterator_test.go:11]: New line

File: pkg/graveler/ref/tag_iterator_test.go [hunk 4]
```diff
@@ -5,6 +5,10 @@ import (
 	"testing"
 	"time"
 
+	"github.com/golang/mock/gomock"
+	"github.com/treeverse/lakefs/pkg/kv"
+	"github.com/treeverse/lakefs/pkg/kv/mock"
+
```
Comment (N-o-Z) [inline pkg/graveler/ref/tag_iterator_test.go:11]: New line

File: pkg/kv/kvtest/store.go [hunk 5]
```diff
@@ -462,12 +462,9 @@ func testStoreMissingArgument(t *testing.T, ms MakeStore) {
 	t.Run("Scan", func(t *testing.T) {
 		start := uniqueKey("test-missing-argument")
 		for _, pk := range [][]byte{nil, {}} {
-			it, err := store.Scan(ctx, pk, start)
+			_, err := store.Scan(ctx, pk, start)
 			if !errors.Is(err, kv.ErrMissingPartitionKey) {
-				t.Errorf("Scan using empty partition key - err=%v, expected %s", err, kv.ErrMissingPartitionKey)
-			}
-			if it != nil {
```
Comment (N-o-Z) [inline pkg/kv/kvtest/store.go:469]: Why did we remove this check?

File: pkg/kv/iterators_test.go [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+	"context"
+	"testing"
+
+	"github.com/golang/mock/gomock"
+	"github.com/treeverse/lakefs/pkg/graveler"
+	"github.com/treeverse/lakefs/pkg/kv"
+	"github.com/treeverse/lakefs/pkg/kv/mock"
+)
+
+func TestPartitionIterator_ClosedBehaviour(t *testing.T) {
```
Comment (N-o-Z) [inline pkg/kv/iterators_test.go:13]: If I understand correctly - this test doesn't reproduce the original issue?

**Review by N-o-Z — CHANGES_REQUESTED** [review]
Thanks, see comments

File: pkg/graveler/ref/commit_ordered_iterator_test.go (hunk 2 above)
Comment (nopcoder) [reply pkg/graveler/ref/commit_ordered_iterator_test.go:12]: Thanks - I need to fix this in my editor settings.

File: pkg/kv/kvtest/store.go (hunk 5 above)
Comment (nopcoder) [reply pkg/kv/kvtest/store.go:469]: - Didn't want to check interface value with nil
- Assumed we test only valid protocol - in this case, when we call the function and there is an error, we can't assume anything about the iterator value.

File: pkg/kv/iterators_test.go (hunk 6 above)
Comment (nopcoder) [reply pkg/kv/iterators_test.go:13]: Wanted test what will happen when getting an error inside SeekGE calling to store's Scan func.
Wanted to address the following:
- When calling close we check the flag and not the pointer.
- Close mark that iterator was closed - prevent crash in calling multiple times to Close.
- Err implementation return the iterator's error based on flag.

**Comment by nopcoder** [comment]
> Thanks, see comments

Thanks - I'll add another test to test SeekGE failed in case of PartitionIterator to cover the previous case.

File: pkg/kv/iterators_test.go [hunk 7]
```diff
@@ -27,3 +28,16 @@ func TestPartitionIterator_ClosedBehaviour(t *testing.T) {
...
+	ctx := context.Background()
+	ctrl := gomock.NewController(t)
+	store := mock.NewMockStore(ctrl)
+	var entIt *mock.MockEntriesIterator
+	entItErr := errors.New("failed to scan")
+	store.EXPECT().Scan(ctx, gomock.Any(), gomock.Any()).Return(entIt, entItErr).Times(1)
+
+	it := kv.NewPartitionIterator(ctx, store, (&graveler.StagedEntryData{}).ProtoReflect().Type(), "partitionKey")
+	it.SeekGE([]byte("test"))
+	it.Close() // verify we don't crash after see failedcx
```
Comment (N-o-Z) [inline pkg/kv/iterators_test.go:42]: see failedcx?

Comment (N-o-Z) [reply pkg/kv/iterators_test.go:42]: Ahh I think I understand now, it's German for "the failed seek" 😆

Comment (nopcoder) [reply pkg/kv/iterators_test.go]: fixed the comment

**Review by N-o-Z — APPROVED** [review]
lgtm, thanks!

File: pkg/kv/iterators_test.go (hunk 7 above)
Comment (nopcoder) [reply pkg/kv/iterators_test.go]: my the hands write something, while the head is thinking about the next line.

**Pull request merged**